### PR TITLE
GSoC'17 Learning compact models for object detection

### PR DESCRIPTION
**Pull request by kvmanohar22**
**~~Merge with extra~~**: https://github.com/opencv/opencv_extra/pull/367

**GSoC Project**: Learning compact models for object detection
**Mentor**: Vladimir Tyan
 
Contents of this **PR**:
- Solvers and model definition files for the trained model on ImageNet.
- Training file to train the model
- labels mapper for object detection

[Tutorial link](http://pullrequest.opencv.org/buildbot/export/pr_contrib/1253/docs/d2/da2/tutorial_dnn_objdetect.html).

### CODE REVIEW

**Comment by kvmanohar22** [comment]
What's the default path that I should set for `Caffe_DIR` and `Protobuf_DIR` ?
The build seems to fail because of this.

**Comment by kvmanohar22** [comment]
@Auron-X This [build warning](http://pullrequest.opencv.org/buildbot/builders/precommit-contrib_docs/builds/11121/steps/patch%20size%20opencv_extra/logs/stdio) seems to be inevitable because of the size of the trained model. Is there any way to suppress this warning ?

**Comment by kvmanohar22** [comment]
@Auron-X, any ideas how to fix [this warning](http://pullrequest.opencv.org/buildbot/builders/precommit-contrib_docs/builds/11138/steps/make%20doxygen/logs/warnings%20%281%29)

File: modules/dnn_objdetect/include/opencv2/core_detect.hpp [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+#ifndef _OPENCV_DNN_OBJDETECT_CORE_DETECT_HPP_
+#define _OPENCV_DNN_OBJDETECT_CORE_DETECT_HPP_
+
+#include <vector>
+#include <memory>
+
+#include <opencv2/highgui.hpp>
+#include <opencv2/imgproc.hpp>
+
```
Comment (mshabunin) [inline modules/dnn_objdetect/include/opencv2/core_detect.hpp]: You have to create documentation group for this module, like this:
https://github.com/opencv/opencv_contrib/blob/ece5ef237c3d02b4dd1c3794cad5ca984d791efa/modules/bgsegm/include/opencv2/bgsegm.hpp#L47-L48
And wrap all functions into this group like this:
https://github.com/opencv/opencv_contrib/blob/ece5ef237c3d02b4dd1c3794cad5ca984d791efa/modules/bgsegm/include/opencv2/bgsegm.hpp#L55-L56

More details here: http://docs.opencv.org/3.3.0/d4/db1/tutorial_documentation.html#tutorial_documentation_commands_group

Comment (kvmanohar22) [reply modules/dnn_objdetect/include/opencv2/core_detect.hpp]: Thanks for the review !

**Comment by kvmanohar22** [comment]
And is there any way to overcome the inevitable doc warning related to size of the patch ?

**Comment by alalek** [comment]
No, this warning is a signal for maintainers to review patches more accurate about added data/commits.
Currently you can just ignore this until the future requests from OpenCV team.

**Comment by kurnianggoro** [comment]
@kvmanohar22 if your big-sized file is not required for building process (e.g. trained model), you can upload that file to another place, in my case, i uploaded to my own github account then just put the link in tutorial or other explanation, if it is required for building, you should put in opencv_3rdparty and see this [explanation](opencv_3rdparty). If you have the file in your history (before you upload and then you delete), you should squash your commits to delete the file from history.

**Comment by kvmanohar22** [comment]
@kurnianggoro The file is not required for the building process. Sure, I'll upload the trained model weights to my own github account !

**Comment by kurnianggoro** [comment]
i think you still have the file in your history, you should rewrite the history, also, in opencv_extra the maximum file size was less than 200 KB, so...i think you should also moves your images or just make them smaller (in my case 150KB max)

**Pull request merged**